test/activerecord/cte_test.rb

```diff
@@ -142,4 +142,35 @@ def test_with_when_merging_relations
     assert_equal(1, merged.size)
     assert_equal(123, merged[0].views_count)
   end
+
+  def test_with_when_merging_relations_with_identical_with_names_and_identical_queries
+    most_popular1 = Post.with(most_popular: Post.where("views_count >= 100"))
+    most_popular2 = Post.with(most_popular: Post.where("views_count >= 100"))
+
+    merged = most_popular1.merge(most_popular2).from("most_popular as posts")
+
+    assert_equal posts(:two, :three, :four), merged.to_a
```

Comment:
This still does not guarantee the order and will fail for some adapters I suppose. Maybe you should add ordering `merged = most_popular1.merge(most_popular2).from("most_popular as posts").order(:id)` or try something like:


```suggestion
    expected_posts = Post.where("views_count >= 100")
    merged = most_popular1.merge(most_popular2).from("most_popular as posts")

    assert expected_posts.any?
    assert_equal expected_posts.to_a, merged.to_a
```

Reply:
Fixed by doing `assert_equal foo.sort, bar.sort`, basically.